#5002, #5003 Replace subpath with overall path in schema

COCOLMAN opened the pull request:
I've tried several ways to solve the subpath problem.

I think it is most appropriate to delete the  subpath.

In addition, I added a filter to decorate.

As mentioned in '# 5002'
I intended to expose more explicit names to the browser.

I used overall instead of subpath.
overall includes all paths.
For example
`First/second/third /` is converted to `first-second-third`.

The browser now displays more clear sentences at the top.
(I used a new filter to remove the '-' and changed it to uppercase.)
![2017-03-23 17 03 22](https://cloud.githubusercontent.com/assets/22088047/24284437/96df8138-10ae-11e7-92ac-7e954b27fd22.png)
![2017-03-23 17 03 47](https://cloud.githubusercontent.com/assets/22088047/24284441/9b7bbc98-10ae-11e7-869f-ad04f81ff2c6.png)

I expect to force users to design url more readable.


More details

In the process of solving the problem,
There are additional modifications.

If `/some/` in a url such as` example/{pk}/some` is appropriate to be included as a sub-property of `/example/` if `/some/` is a` custom action`.


However, if `/some/` is a method that is not a `custom action`(Having different 'actions'), I do not think it is appropriate to include it as a subproperty in `/example/`.

So in this case, instead of being a sub-property of `example`, it has a new overall of` example-some`.

My opinion is too much.
I am afraid that this is against the overall design.

COCOLMAN commented:
sample image is a change of url and path parameters to help you understand.
Do not mind why have `name` and` password`.

tomchristie commented:
We can't really accept this change as it stands, for two reasons:

* It'd change the layout, breaking API clients that folks may be using.
* It's not sufficiently clear from the description exactly which layouts would change and which would not (Would need a fairly comprehensive example)

It's awkward that REST framework's schema generation can't always adequately automatically determine a good layout, but I think that for now it's probably just a constraint that we need to live with, and guide folks towards how they can explicitly declare a schema in cases where that's needed.

COCOLMAN commented:
If you look at # 5003, you can see that this feature is not working properly.

To solve this problem, I felt that many of the methods associated with the 'get_links' method needed to be redesigned.

Because there are too many modifications, I need a lot of feedback. I hope to have many conversations.

You say   

> It's not quite clear from the description exactly which layouts would change and which would not (Would need a fairly comprehensive example)


But I can not explain the changes to the layout.

Since the layout did not work before then

COCOLMAN commented:
When url is a tree structure,
```
##example urls##

# root/urls.py
urlpatterns = [
    url(r'^admin/', admin.site.urls),
    url(r'^member/', include('member.urls')),
    url(r'^blog/', include('post.urls')),
    url(r'^docs/', include_docs_urls(title='My API TITLE'))
]

# post/urls.py
urlpatterns = [
    url('^post/$', views.PostListCreateView.as_view())
]

# member/urls.py
urlpatterns = [
    url(r'^signup/$', views.UserCreate.as_view()),
    url(r'^myinfo/(?P<pk>[0-9]+)$', views.TestView.as_view())
]
```

The schema created dict that contains dict.

Issue # 5003
This happens because there is no 'action' at the moment API Document expects to have an 'action' in the schema.


The schema for these URLs is:

```
<My API TITLE "http://127.0.0.1:8000/docs/schema.js">
    blog: {
        post: {
            list()
            create(title, author, [content])
        }
    }
    member: {
        myinfo: {
            read(id)
            update(id, name, password)
            partial_update(id, [name], [password])
        }
        signup: {
            create(name, password)
        }
    }
```

> Removing `include ('post.urls')` from `root / urls` works fine because all urls remove the common` member / `and the schema looks like this:
```
<My API TITLE "http://127.0.0.1:8000/docs/schema.js">
    myinfo: {
        read(id)
        update(id, name, password)
        partial_update(id, [name], [password])
    }
    signup: {
        create(name, password)
    }
```


In order to make this schema work, I changed it as follows.

```
<My API TITLE "http://127.0.0.1:8000/docs/schema.js">
    blog-post: {
        list()
        create(title, author, [content])
    }

    member-myinfo: {
        read(id)
        update(id, name, password)
        partial_update(id, [name], [password])
    }
    member-signup: {
        create(name, password)
    }
```

I Include all url paths to avoid losing meaning.

URLs are linked with '-' so you can use coreapi normally from the command line.


The above content is the key content I have used to solve this problem.


There are two additional changes in the above 'PR'.


### 1
```
In addition, I added a filter to decorate.
...
(I used a new filter to remove the '-' and changed it to uppercase.)
```

### 2
```
If /some/ in a url such asexample/{pk}/some is appropriate to be included as a sub-property of /example/ if /some/ is acustom action.

However, if /some/ is a method that is not a custom action(Having different 'actions'), I do not think it is appropriate to include it as a subproperty in /example/.

So in this case, instead of being a sub-property of example, it has a new overall ofexample-some.
```

I want feedback.

COCOLMAN commented:
What do you think about key content except for additional 1, 2?

tomchristie commented:
The schema nested by more than two deep looks okay to me - it's not clear what squishing that down to two levels, using hyphenation achieves. If we've any problem with rendering 3 deep schemas in our docs then we should fix that, but the layout looks okay to me as it stands.

(Also thanks - discussing this from a "I think X layout, should be Y instead" is much clearer.

COCOLMAN commented:
The schema nested by more than 2 deep has problem. 
https://github.com/encode/django-rest-framework/issues/5003

Am I using it wrong?
